client/src/main/scala/com/redi/scalademo/presentation/FormListener.scala

```diff
@@ -95,34 +75,66 @@ class FormListener(
           formElement.find(elementNamed(name)).addClass(DangerClass)
         }
       }
-      name match {
-        case Augend ⇒
-          augend = value
-        case Addend ⇒
-          addend = value
+    }
+    allFormValuesAreValid
+  }
+
+  private def serializedFormData: js.Array[js.Dictionary[String]] = {
+    formElement.serializeArray().asInstanceOf[js.Array[js.Dictionary[String]]]
+  }
+
+  private def calculableFormData: js.Array[js.Dictionary[String]] = {
+    var calculableFormData = js.Array[js.Dictionary[String]]()
+    for (formControlData: js.Dictionary[String] ← serializedFormData) {
+      if (formControlData(Name) != Validate) {
+        calculableFormData += formControlData
+      }
+    }
+    calculableFormData
+  }
+
+  private def shouldValidate: Boolean = {
+    var shouldValidate = false
+    for (formControlData: js.Dictionary[String] ← serializedFormData) {
+      if (formControlData(Name) == Validate) {
+        shouldValidate &= formControlData(Value).nonEmpty
       }
     }
+    shouldValidate
   }
 
   private def invokeAddition(): Unit = {
-    if (allFormValuesAreValid) {
-      val futureResult: Future[ValidatedResult[BigDecimal]] =
-        client[Calculator].add(augend, addend).call()
-      for (result: ValidatedResult[BigDecimal] ← futureResult) {
-        result match {
-          case Left(validationFailures) ⇒
-            for (validationFailure: ValidationFailure ← validationFailures) {
-              formElement
-                .find(elementNamed(validationFailure.formControlName))
-                .addClass(DangerClass)
-            }
-          case Right(summand: BigDecimal) ⇒
-            summandElement
-              .value(summand.toString)
-              .addClass(InfoClass)
-        }
+    val futureResult: Future[ValidatedResult[BigDecimal]] =
+      client[Calculator].add(augend, addend).call()
```

Comment:
Trying to put myself in the mindset of NY folks, here. Before, getting `augend` and `addend` required only one iteration through `calculableFormData` in `validateAndSetStyles()` (formerly `validateAndExtractNumbers()`). I think I'd like to retain the efficiency, small as though it may be. Perhaps return a tuple or a special-purpose nested class instance?